Loading & Inspecting Data

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split


In [168]:
#Loading the dataset
path = "/content/Telco-Customer-Churn.csv"
df = pd.read_csv(path)
df.head()

,Unnamed: 0,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


##EXPLORATORY DATA ANALYSIS (EDA)

In [169]:
# showing the lists of the columns
df.columns.tolist()

['Unnamed: 0',
 'CustomerID',
 'Count',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Lat Long',
 'Latitude',
 'Longitude',
 'Gender',
 'Senior Citizen',
 'Partner',
 'Dependents',
 'Tenure Months',
 'Phone Service',
 'Multiple Lines',
 'Internet Service',
 'Online Security',
 'Online Backup',
 'Device Protection',
 'Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charges',
 'Total Charges',
 'Churn Label',
 'Churn Value',
 'Churn Score',
 'CLTV',
 'Churn Reason']

In [170]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 34 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         7043 non-null   int64  
 1   CustomerID         7043 non-null   object 
 2   Count              7043 non-null   int64  
 3   Country            7043 non-null   object 
 4   State              7043 non-null   object 
 5   City               7043 non-null   object 
 6   Zip Code           7043 non-null   int64  
 7   Lat Long           7043 non-null   object 
 8   Latitude           7043 non-null   float64
 9   Longitude          7043 non-null   float64
 10  Gender             7043 non-null   object 
 11  Senior Citizen     7043 non-null   object 
 12  Partner            7043 non-null   object 
 13  Dependents         7043 non-null   object 
 14  Tenure Months      7043 non-null   int64  
 15  Phone Service      7043 non-null   object 
 16  Multiple Lines     7043 

From the info, Total Charges column is an object but it takes a numeric data, so we will need to convert it to a numeric column so that it doesn't affect us  later.


In [171]:
# Changing the Total Charges column to a numeric column because it was originally an object dtype.
df["Total Charges"] = pd.to_numeric(df["Total Charges"], errors="coerce")



In [172]:
df.isnull().sum()

,0
Unnamed: 0,0
CustomerID,0
Count,0
Country,0
State,0
City,0
Zip Code,0
Lat Long,0
Latitude,0
Longitude,0


The total Charges Column now reflects 11 missing values after converting it to a numeric column while other columns do not have any missing value.

In [173]:
# Filling the missing values in total Charges Column.
df["Total Charges"].fillna(df["Total Charges"].median(), inplace=True)

/tmp/ipykernel_967/3130141094.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Total Charges"].fillna(df["Total Charges"].median(), inplace=True)


In [174]:
# checking to confirm all the missing values have been filled.
df["Total Charges"].isnull().sum()

np.int64(0)

All the missing Values have now being filled.

CONVERTING CHURN

In [175]:
# converting the target (Churn)
df["Churn Label"]= df["Churn Label"].map({"Yes": 1, "No" : 0})

In [176]:
# checking the Churn counts
df["Churn Label"].value_counts()

,count
Churn Label,
0,5174
1,1869


#CALCULATING CHURN RATE

In [177]:
df.columns

Index(['Unnamed: 0', 'CustomerID', 'Count', 'Country', 'State', 'City',
       'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Gender',
       'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months',
       'Phone Service', 'Multiple Lines', 'Internet Service',
       'Online Security', 'Online Backup', 'Device Protection', 'Tech Support',
       'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing',
       'Payment Method', 'Monthly Charges', 'Total Charges', 'Churn Label',
       'Churn Value', 'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='object')

In [178]:
df["Churn"]= df["Churn Label"]
Churn_Rate = df["Churn"].mean()
print(Churn_Rate)

0.2653698707936959


In [179]:
df.columns

Index(['Unnamed: 0', 'CustomerID', 'Count', 'Country', 'State', 'City',
       'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Gender',
       'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months',
       'Phone Service', 'Multiple Lines', 'Internet Service',
       'Online Security', 'Online Backup', 'Device Protection', 'Tech Support',
       'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing',
       'Payment Method', 'Monthly Charges', 'Total Charges', 'Churn Label',
       'Churn Value', 'Churn Score', 'CLTV', 'Churn Reason', 'Churn'],
      dtype='object')

The above shows that the churn rate is 26% which means that;


*   Out of every **100** customers **26** leave



*   That's more than **1** in **4** **Customers**

*   loosing 1 in 4 customers shows that there is a ***major revenue leakage***






If the Churn rate is this high, it means there is a retention problem and we

need to know which customers are leaving and why.





#Revenue KPI

In [180]:
avg_revenue = df["Monthly Charges"].mean()
total_revenue = df["Monthly Charges"].sum()

print(avg_revenue, total_revenue)

64.76169246059918 456116.6


#Churn By Tenure Group

In [181]:
# creating Tenure group
df["tenure_group"] = pd.cut(df["Tenure Months"],
bins = [0,12,24,48,72],
labels = ['0-1yr','1-2yr','2-4yr','4-6yr'])

#calculating churn by tenure group
df.groupby("tenure_group")["Churn"].mean()

/tmp/ipykernel_967/3220778613.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby("tenure_group")["Churn"].mean()


,Churn
tenure_group,
0-1yr,0.476782
1-2yr,0.287109
2-4yr,0.203890
4-6yr,0.095132


**Insight:**


*  Customers in the **0–1 year** group have **~48%** churn (very high)
*  Customers in **4–6 years** have **~9.5%** churn (very low)


**Business Interpretation:**

*  Early-stage customers are at highest risk
*  Long-term customers are stable and valuable


**Recommendation:**


*  Improve onboarding and early retention strategies
*  Offer incentives for new customers








#Churn By Contract Type

In [182]:
df.groupby('Contract')['Churn'].mean()

,Churn
Contract,
Month-to-month,0.427097
One year,0.112695
Two year,0.028319


**INSIGHT:**

Customers on month-to-month contracts have significantly higher churn (~43%)

 compared to those on long-term contracts (~3%), indicating that contract structure strongly influences retention.”

 **Business Interpretation**:

 **Problem**:

Flexible (monthly) customers leave easily

High churn = unstable revenue

**Opportunity**:

Long-term contracts drastically reduce churn

More predictable revenue

**BUSINESS ACTIONS/ RECCOMENDATION:**

**Strategy 1:**

Incentivise long-term contracts

Offer discounts for 1-year / 2-year plans

Bundle benefits for long-term commitment

**Strategy 2**:

Adjust pricing for risk

Month-to-month = higher price (risk premium)

Long-term = lower price (reward loyalty)

**Strategy 3**:

Target high-risk customers

Identify monthly customers early

Offer conversion incentives


#Risk + Revenue


##Creating Value segments

In [183]:
# Creating Value segments
df['Value_segments'] = pd.qcut(df['Monthly Charges'], 3, labels = ['low','Medium','High'])
df['Value_segments']

,Value_segments
0,Medium
1,Medium
2,High
3,High
4,High
...,...
7038,low
7039,High
7040,High
7041,low


##Analysing Revenue + Churn

In [184]:
#
df.groupby('Value_segments')[['Monthly Charges','Churn']].mean()


/tmp/ipykernel_967/59426792.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('Value_segments')[['Monthly Charges','Churn']].mean()


,Monthly Charges,Churn
Value_segments,,
low,27.988622,0.158656
Medium,69.036333,0.296802
High,97.326438,0.340861


The result above shows that;

**High-value customers** pay the most **(~£97)** but also have the highest

Churn **(~£34)**

This is VERY dangerous for the business



**Low-value customers** Pay less **(~£28)** and they have much lower

Churn **(~£16%)**

They are Less valuable, but more stable.


 **INSIGHT:**

 “Higher-value customers have significantly higher churn, indicating that the most profitable segment is also the most at risk.”

 **BUSINESS INTERPRETATION:**

 **Problem**

 The customers generating the most revenue are Leaving more often and
  Causing major revenue loss.

**Risk**

Losing high-value customers = losing a large portion of revenue


**BUSINESS STRATEGY:**

**Strategy 1:** Protect high-value customers by giving them Loyalty rewards,

Personalised offers and better customer experience.

**Strategy 2:** Investigate WHY they churn and find out if it is due to  
 Pricing too high,Poor perceived value or Better competitor offers.

**Strategy 3:** Adopt Smart pricing by offering customers value based pricing

 and retaining them strategically rather than charging them blindly.










  **SUMMARY**

**OVERALL PROBLEM:** Churn Rate= **26%(High)**

**WHERE CHURN HAPPENS:** New customers → **~48% churn**

**WHY IT HAPPENS:**
Flexible contracts → **~43% churn**

**WHERE MONEY IS AT RISK:**
High-value customers → **34% churn**

In [185]:
# creating a copy for tableau dashboard
tableau_df = df.copy()


#MODELLING

In [186]:

# Dropping CustomerID Column
model_df = df.drop(['CustomerID','Churn Score','Churn Reason','Churn Value','Churn','Churn Label','Country','State','City','Lat Long','Unnamed: 0'], axis = 1)

#Converting Categorical Columns to Numeric
model_df = pd.get_dummies(model_df, drop_first = True)

#Defining Features and Target
x = model_df
y = df['Churn']

#Splitting the Data
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.2,random_state=42)

#Training the Model
Model = LogisticRegression(max_iter = 1000)
Model.fit(x_train,y_train)
Model.score(x_train,y_train)

#Creating Churn Risk Score
tableau_df["Risk_score"] = Model.predict_proba(x)[:, 1]
tableau_df["Risk_score"].head()

#previewing the risk score
tableau_df[['Risk_score','Churn']].head()



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


,Risk_score,Churn
0,0.334433,1
1,0.450295,1
2,0.578814,1
3,0.295579,1
4,0.191818,1


with the risk score above, we can now predict which customers are likely to churn and take a proactive measure before they leave.

#Creating Risk + Value

##Creating Risk Segment

In [188]:
tableau_df['Risk_segment']= pd.cut(tableau_df['Risk_score'], 3,
                    labels = ['Low','Medium','High'])
tableau_df[['Risk_score','Risk_segment']]

,Risk_score,Risk_segment
0,0.334433,Medium
1,0.450295,Medium
2,0.578814,Medium
3,0.295579,Low
4,0.191818,Low
...,...,...
7038,0.004650,Low
7039,0.044963,Low
7040,0.063426,Low
7041,0.205121,Low


##Analysing Risk VS Value

In [189]:
tableau_df.groupby(['Risk_segment','Value_segments']).size()


/tmp/ipykernel_967/426662710.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  tableau_df.groupby(['Risk_segment','Value_segments']).size()


Risk_segment  Value_segments
Low           low               1897
              Medium            1397
              High              1218
Medium        low                358
              Medium             473
              High               633
High          low                 96
              Medium             475
              High               496
dtype: int64

The risk vs Value Analysis shows that the largest group of high-risk customers is:

**High Risk + High Value = 496 customers**

This means the customers who pays the most money and are most likely to leave.

#**Business Interpretation:**

**Problem:** The most valuable customers are also the most at risk

**Risk:** If these customers churn,there will be huge revenue loss and

Profit will drop significantly.







##**FINAL PRICING STRATEGY RECOMMENDATION**

**1. High Value + High Risk (MOST IMPORTANT)**




*   Prioritise retention
*   Offer incentives

*  Personalised engagement

 **2.High Value + Low Risk**



*   Maintain pricing
*  Upsell / cross-sell

**3. Low Value + High Risk**



*   Increase pricing OR
*   Limit investment


**4. Low Risk customers**


*   Stable → minimal intervention


















##**FINAL SUMMARY**

I analysed customer churn and found that early-stage customers and those on flexible contracts have significantly higher churn. More importantly, high-value customers also exhibit higher churn rates, which presents a major revenue risk. By segmenting customers based on risk and value, I identified that high-value, high-risk customers should be prioritised for retention strategies, while pricing adjustments can be applied to lower-value, higher-risk segments.

In [190]:
tableau_df.to_excel("tableau_telco_dashboard_data.xlsx", index=False)
from google.colab import files
files.download("tableau_telco_dashboard_data.xlsx")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>